lib/srv/alpnproxy/helpers_test.go

```diff
@@ -0,0 +1,217 @@
+	require.NoError(t, err)
+
+	syncC := make(chan struct{})
+	go func() {
+		close(syncC)
+		err := svr.Serve(context.Background())
+		require.NoError(t, err)
+	}()
+	<-syncC
+
+	t.Cleanup(func() {
+		svr.Close()
+	})
+}
+
+func mustGenSelfSignedCert(t *testing.T) *tlsca.CertAuthority {
+	caKey, caCert, err := tlsca.GenerateSelfSignedCA(pkix.Name{
+		CommonName: "localhost",
+	}, []string{"localhost"}, defaults.CATTL)
+	require.NoError(t, err)
+
+	ca, err := tlsca.FromKeys(caCert, caKey)
+	require.NoError(t, err)
+	return ca
+}
+
+type signOptions struct {
+	identity tlsca.Identity
+}
+
+func withIdentity(identity tlsca.Identity) signOptionsFunc {
+	return func(o *signOptions) {
+		o.identity = identity
+	}
+}
+
+type signOptionsFunc func(o *signOptions)
+
+func mustGenCertSignedWithCA(t *testing.T, ca *tlsca.CertAuthority, opts ...signOptionsFunc) tls.Certificate {
+	options := signOptions{
+		identity: tlsca.Identity{Username: "test-user"},
+	}
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	subj, err := options.identity.Subject()
+	require.NoError(t, err)
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	clock := clockwork.NewRealClock()
+	tlsCert, err := ca.GenerateCertificate(tlsca.CertificateRequest{
+		Clock:     clock,
+		PublicKey: privateKey.Public(),
+		Subject:   subj,
+		NotAfter:  clock.Now().UTC().Add(time.Minute),
+		DNSNames:  []string{"localhost", "*.localhost"},
+	})
+	require.NoError(t, err)
+
+	keyRaw := x509.MarshalPKCS1PrivateKey(privateKey)
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: keyRaw})
+	cert, err := tls.X509KeyPair(tlsCert, keyPEM)
+	require.NoError(t, err)
+	return cert
+}
+
+func mustReadFromConnection(t *testing.T, conn net.Conn, want string) {
+	buff, err := io.ReadAll(conn)
+	require.NoError(t, err)
+	require.Equal(t, string(buff), want)
+}
+
+func mustCloseConnection(t *testing.T, conn net.Conn) {
+	err := conn.Close()
+	require.NoError(t, err)
+}
+
+func mustCreateLocalListener(t *testing.T) net.Listener {
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		l.Close()
+	})
+	return l
+}
+
+func mustSuccessfullyCallHTTPSServer(t *testing.T, addr string, client http.Client) {
+	resp, err := client.Get(fmt.Sprintf("https://%s", addr))
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+}
+
+func mustStartHTTPServer(t *testing.T, l net.Listener) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {})
+	go http.Serve(l, mux)
+}
+
+func mustStartLocalProxy(t *testing.T, config LocalProxyConfig) {
+	lp, err := NewLocalProxy(config)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		err = lp.Close()
+		require.NoError(t, err)
+	})
+	syncC := make(chan struct{})
+	go func() {
+		close(syncC)
+		err := lp.Start(context.Background())
+		require.NoError(t, err)
+	}()
+	<-syncC
```

Comment:
Yeah maybe I am missing something but I'm not sure what `syncC` is doing other than waiting until the goroutine at least starts? I don't think it's a guarantee that the proxy has started though